export storageconsumer data as metrics and alerts

leelavg opened the pull request:
- encode the versions from storageconsumer CR status to comparable numbers
- export these numbers are metrics
- create alerts for version incompatbility and client checkin via heartbeats

leelavg commented:
only last commit is reviewable for now.

leelavg commented:
@aruniiird, @jmolmo could you pls review last commit in this PR?

jmolmo commented on metrics/internal/collectors/storageconsumer.go:
```diff
@@ -59,12 +114,84 @@ func (c *StorageConsumerCollector) Run(stopCh <-chan struct{}) {
 			string(storageConsumer.Status.State))
+
+		ch <- prometheus.MustNewConstMetric(c.LastHeartbeat,
+			prometheus.GaugeValue, float64(storageConsumer.Status.LastHeartbeat.Time.Unix()),
```
Probably a heartbeat based on a timestamp would be of type "counter" instead "gauge". Do you expect to have decrements on this metric?

leelavg replied:
There'll be no decrements, however _count and _sum derived metrics from a counter type doesn't mean anything in this context and so I used Gauge to represent the instant value with no correlation to previous values.

jmolmo commented on metrics/internal/collectors/storageconsumer.go:
```diff
@@ -49,6 +99,11 @@ func (c *StorageConsumerCollector) Collect(ch chan<- prometheus.Metric) {
 	ds := []*prometheus.Desc{
 		c.StorageConsumerMetadata,
+		c.LastHeartbeat,
+		c.ProviderClusterVersion,
```
I think that all the "*Version" metrics should be manged as labels of LastHeartBeat metric. Why do you want to have them as metrics instead of labels?

jmolmo commented on metrics/mixin/alerts/storage-client.libsonnet:
```diff
@@ -0,0 +1,69 @@
+          {
+            alert: 'StorageClientHeartbeatWarn',
+            expr: |||
+              time() - %(clientCheckinWarnSec)d > ocs_storage_client_last_heartbeat > 0
```
Using "()" too enclose comparations probably the expression is more readable.

jmolmo commented on metrics/mixin/alerts/storage-client.libsonnet:
```diff
@@ -0,0 +1,69 @@
+            # divide by 1000 here removes patch version
+            alert: 'StorageClientIncompatibleCluster',
+            expr: |||
+              floor(abs((ocs_storage_client_cluster_version > 0) - ignoring(storage_consumer_name) group_left() (ocs_storage_provider_cluster_version > 0)) / 1000) > %(clientClusterMinorVerDiff)d
```
do you really want
floor(abs((ocs_storage_client_cluster_version > 0)
instead of:
floor(abs(ocs_storage_client_cluster_version) > 0)

jmolmo commented on metrics/mixin/alerts/storage-client.libsonnet:
```diff
@@ -0,0 +1,69 @@
+              severity: 'critical',
+            },
+            annotations: {
+              message: 'Storage Client ({{ $labels.storage_consumer_name }}) heartbeat missed for more than %d (s)' % $._config.clientCheckinWarnSec,
```
I think you want here to use `clientChekinCritSecs`

leelavg replied:
Yep, missed it.

jmolmo reviewed (changes requested):
Probably you need to replace the var in the "critical" alert

Please add documentation about the new alerts in the runbooks repo.
Here you have an example for [CephClusterCriticallyFull alert](https://github.com/openshift/runbooks/blob/master/alerts/openshift-container-storage-operator/CephClusterCriticallyFull.md)

Besides that, Alerts in OpenShift will ship a link to the corresponding runbook in this repository, to make fixing problems even easier. Link should be passed as runbook_url field in alert annotations
See the [example](https://github.com/openshift/cluster-monitoring-operator/blob/1959ea9461c47eee506a497b21f10b0e2c3701ac/assets/cluster-monitoring-operator/prometheus-rule.yaml#L31-L40)

leelavg commented:
> Please add documentation about the new alerts in the runbooks repo.
- @jmolmo, can this be deferred to a later point, ie, after 4.14 and before 4.14.z as we require the PR to be feature complate
- And also we aren't merging to `main` as this will be a custom build until release and I'll definitely raise PRs for the runbooks after this PR gets merged as I don't want to update at two places in parallel if we require changes to the alerts wrt pr comments

umangachapagain reviewed (changes requested):
Lots of unrelated changes in the alert yaml files, please remove it from commit.
Looks good otherwise.

leelavg commented:
> Lots of unrelated changes in the alert yaml files, please remove it from commit. Looks good otherwise.

- I just followed README and these are generated when I ran corresponding `make` targets
- #2225 should probably fix it, anyways, if these are generated during build I can remove from this PR, can I?

jmolmo commented:
> > Lots of unrelated changes in the alert yaml files, please remove it from commit. Looks good otherwise.
> 
> * I just followed README and these are generated when I ran corresponding `make` targets
> * [Ran 'make gen-latest-prometheus-rules-yamls' command #2225](https://github.com/red-hat-storage/ocs-operator/pull/2225) should probably fix it, anyways, if these are generated during build I can remove from this PR, can I?

I think, everything will be more clear if we merge first:
https://github.com/red-hat-storage/ocs-operator/pull/2225

@umangachapagain can you review it and approve if it is ok?

leelavg commented:
@jmolmo both the branches differ and for now if they are generated during build time then this PR isn't dependent on any other.

jmolmo commented:
> @jmolmo both the branches differ and for now if they are generated during build time then this PR isn't dependent on any other.

It seems that the modifications in the rules file are not generated during the build, you need to execute explicitly the command "make gen-latest-prometheus-rules-yamls":
#2225 is set because it was forgotten to execute it.
If you want to add your rules, you will need also to execute  the make command an you will generate your modifications and also the modifications in #2225.
This is why I say that in order to make thing easier for ourselves in the future, probably #2225 should be merged first.

leelavg commented:
> It seems that the modifications in the rules file are not generated during the build
- ack, got it. Here we are in a different situation, #2225 was raised for a different branch, for the HCI effort pls consider the current base branch (`fusion-hci-4.14`) is entirely different
- so I added the changes generated after following README to this PR itself
- when we take in changes from main, these changes shouldn't conflict (mostly) as they are generated by same tooling
---
- I've rebased on the base branch and this PR can receive your votes for the merge if you are fine with it @jmolmo @umangachapagain 

thanks.

umangachapagain commented on metrics/deploy/prometheus-ocs-rules.yaml:
```diff
@@ -224,3 +227,41 @@ spec:
         severity: critical
+  - name: storage-client-alerts.rules
+    rules:
+    - alert: StorageClientHeartbeatWarn
```
```suggestion
    - alert: StorageClientHeartbeatMissed
```

leelavg replied:
can we have two alerts with same name?

leelavg replied:
done.

umangachapagain commented on metrics/deploy/prometheus-ocs-rules.yaml:
```diff
@@ -224,3 +227,41 @@ spec:
+        (time() - 120) > (ocs_storage_client_last_heartbeat > 0)
+      labels:
+        severity: warning
+    - alert: StorageClientHeartbeatCrit
```
```suggestion
    - alert: StorageClientHeartbeatMissed
```

umangachapagain commented on metrics/deploy/prometheus-ocs-rules.yaml:
```diff
@@ -224,3 +227,41 @@ spec:
+        (time() - 300) > (ocs_storage_client_last_heartbeat > 0)
+      labels:
+        severity: critical
+    - alert: StorageClientIncompatiblePlatform
```
```suggestion
    - alert: StorageClientIncompatiblePlatformVersion
```

leelavg replied:
done.

umangachapagain commented on metrics/deploy/prometheus-ocs-rules.yaml:
```diff
@@ -224,3 +227,41 @@ spec:
+        floor((ocs_storage_client_cluster_version > 0) - ignoring(storage_consumer_name) group_left() (ocs_storage_provider_cluster_version > 0) / 1000) > 1 <= -1
+      labels:
+        severity: warning
+    - alert: StorageClientIncompatibleOperator
```
```suggestion
    - alert: StorageClientIncompatibleOperatorVersion
```

leelavg replied:
done.

umangachapagain commented on metrics/deploy/rbac.yaml:
```diff
@@ -73,6 +73,14 @@ rules:
     - get
     - list
     - watch
+- apiGroups:
```
these RBAC changes should also be added to https://github.com/red-hat-storage/ocs-operator/blob/2f5040f0050e8a33c3ae4e6c44570133cfb8cdc6/config/rbac/exporter-role.yaml for it to be applied to the cluster.

leelavg replied:
done.

umangachapagain commented on metrics/deploy/rbac.yaml:
```diff
@@ -73,6 +73,14 @@ rules:
     - watch
+- apiGroups:
+  - config.openshift.io
+  resources:
+  - clusterversions
+  verbs:
+    - get
+    - list
+    - watch
```
These are still not added to the other file.

leelavg replied:
I don't think so, already added to exporter-role.yaml
